destination/clickhouse.go

```diff
@@ -30,146 +31,185 @@ type ClickHouseConnection struct {
 }
 
 func GetClickHouseConnection(ctx context.Context, configuration map[string]string) (*ClickHouseConnection, error) {
-	hostname := fmt.Sprintf("%s:%s",
-		GetWithDefault(configuration, "hostname", "localhost"),
-		GetWithDefault(configuration, "port", "9000"))
-	username := GetWithDefault(configuration, "username", "default")
-	password := GetWithDefault(configuration, "password", "")
-	database := GetWithDefault(configuration, "database", "default")
-	options := &clickhouse.Options{
-		Addr: []string{hostname},
-		Auth: clickhouse.Auth{
-			Username: username,
-			Password: password,
-			Database: database,
-		},
-		Protocol: clickhouse.Native,
-		Settings: clickhouse.Settings{
-			"allow_experimental_object_type": 1,
-			"date_time_input_format":         "best_effort",
-			//"async_insert":                   1,
-			//"wait_for_async_insert":          1,
-		},
-		ClientInfo: clickhouse.ClientInfo{
-			Products: []struct {
-				Name    string
-				Version string
-			}{
-				{Name: "fivetran-destination", Version: Version},
+	return RetryNetErrorWithData(func() (*ClickHouseConnection, error) {
+		hostname := fmt.Sprintf("%s:%s",
+			GetWithDefault(configuration, "hostname", "localhost"),
+			GetWithDefault(configuration, "port", "9000"))
+		username := GetWithDefault(configuration, "username", "default")
+		password := GetWithDefault(configuration, "password", "")
+		database := GetWithDefault(configuration, "database", "default")
+		options := &clickhouse.Options{
+			Addr: []string{hostname},
```

Comment:
what if I run on-premise with multiple nodes?

Reply:
It's still in the TODO list. https://github.com/ClickHouse/clickhouse-fivetran-destination/blob/efe89020663038a90664a4a3d3954d1ad794156c/destination/clickhouse.go#L25-L31
We need to ask macro values

```xml
<clickhouse>
  <macros>
    <cluster>test_cluster</cluster>
    <replica>clickhouse1</replica>
    <shard>1</shard>
  </macros>
</clickhouse>
```

to generate something like

```sql
CREATE TABLE foo ON CLUSTER '{cluster}'
(id UInt64, name String, ...)
ENGINE ReplicatedMergeTree(
  '/clickhouse/{cluster}/tables/{database}/{table}/{shard}',
  '{replica}'
)
```

Reply:
Will add as a follow-up, see #3